src/rootfstreemanager.cc

```diff
@@ -1,15 +1,34 @@
 #include "rootfstreemanager.h"
+#include "ostree/repo.h"
 
 bool RootfsTreeManager::fetchTarget(const Uptane::Target &target, Uptane::Fetcher &fetcher, const KeyManager &keys,
                                     const FetcherProgressCb &progress_cb, const api::FlowControlToken *token) {
+  (void)fetcher;
+
   if (!target.IsOstree()) {
     throw std::logic_error("Unsupported type of Target, only OSTREE Target is expected");
   }
 
   RequestHeaders headers;
-  getPullHeaders(headers);
+  std::string serverUrl = getServerUrl(headers);
+  std::string gcsRemote{"gcs"};
+
+  // If the server URL is not available then use the default ostree server specified in sota.toml
+  if (!serverUrl.empty()) {
+    setRemote(gcsRemote, serverUrl);
+  }
 
-  return OstreeManager::pull(config.sysroot, config.ostree_server, keys, target, token, progress_cb, headers).success;
+  return OstreeManager::pull(config.sysroot, config.ostree_server, keys, target, token, progress_cb,
+                             serverUrl.empty() ? nullptr : gcsRemote.c_str(), headers)
+      .success;
 }
 
-void RootfsTreeManager::getPullHeaders(RequestHeaders &headers) {}
+std::string RootfsTreeManager::getServerUrl(RequestHeaders &auth_header) {
+  (void)auth_header;
+  return "";
```

Comment:
seems sane so far. I guess the next step will be deciding how to fill this logic in?

Reply:
yep. I tested it locally by hard-coding the bucket URL and oauth token, works as expected.
The next step is to add a new endpoint to device-gateway or ostreeproxy that will provide an oauth token and download URL(s) to devices.

Reply:
I think it makes sense to fallback to downloading via ostreeproxy in any case.